Fixed issues with carolina

agchitu opened the pull request:
After debugging in the context of Everest a number of issues were identified and fixed. 
Some of the issues were introduced when taking out the mpi dependency. The current version works with latest version of  libseba (which is still to be pushed to its repository) and has been tested on U14.04 with python 2.7.6

Commit:
After debuging fixed the code to work properly:
   -  set the analysis_components as expected
   -  store and use the driver_instance

Commit:
Removed hard coded problem definition from DakotaInput
The problem definition should be provided in the calling code and passed to the DakotaBase
constructor. If None is provided an exception will be raised.
The DakotaInput is closely linked with the dakota version targeted and currently does not yet
have generic implementation

joakim-hove commented on src/dakota.py:
```diff
@@ -32,25 +32,48 @@
         """
-        Write `self.input` to `infile`, providing `self` as `data`.
+        Write `self.input` to `infile`, providing `self` as `driver_instance` if requested.
+        If additional model instance is provided pass that as 'driver_instance'
```
This was less than clear.

agchitu replied:
I'll clean this up, also taking into account one of your comments below.

joakim-hove commented on src/dakota.py:
```diff
@@ -32,25 +32,48 @@
         DAKOTA will then call our :meth:`dakota_callback` during the run.
         """
-        self.input.write_input(infile, data=self, other_data=None)
+        if self_driver and other_model_instance is not None:
```
This seems like a error condition - raise an exception!

joakim-hove commented on src/dakota.py:
```diff
@@ -32,25 +32,48 @@
+        if self_driver:
+            self.input.write_input(infile, driver_instance=self)
+
+        elif other_model_instance is not None:
```
If the user for some reason want to use "another driver" - whatever that means, she should instantiate another `DakotaBase` instance - this was a really confusing mix of instance based and classmethod-like behavior.

agchitu replied:
I agree with this very much... it was in the original code and I though it might be useful at some point, but indeed I would never use it myself. Seeing you are thinking the same I will remove this possibility and only allow for an instance derived from DakotaBase to be used.

joakim-hove commented on src/dakota.py:
```diff
@@ -104,41 +127,37 @@ def __init__(self, **kwargs):
-                if section == 'interface' and data is not None:
+                # Write the driver instance id as analysis_components
+                # if driver instance is specified and store the reverence to this instance for later use
+                if section == 'interface' and driver_instance is not None:
```
Why of why are we fiddling with this driver management?

agchitu replied:
It can be that you still want to use an external model (external application) as a driver. This is the whole reason for this checks here. I do not know if we want to still keep this in carolina solution or make it Python only. As it is right now the callback function is not ready to run anything but the driver anyway. If you say you would rather see a clean, fit for purpose solution I will clean all this up now.

Commit:
Cleaned up the code and created a fit for purpose version of the
dakota module. This means that only a Python based interface with
dakota_callback as a callback function and a driver instance of a
child class for DakotaBase is allowed.
Checks have been put in place to guard for the way this module
is expected to be used and exceptions are raised.
Docstrings for classes and methods are provided.
